### PR TITLE
Build webnn external delegate in native

### PR DESCRIPTION
**Pull request by Honry**

### CODE REVIEW

File: tensorflow/lite/delegates/webnn/BUILD [hunk 1]
```diff
@@ -31,10 +31,28 @@ cc_library(
...
+        "@webnn_native_project//:webnn-native-1",
         "@FP16",
     ],
 )
 
+cc_binary(
+    name = "webnn_external_delegate.so",
+    srcs = [
+        "webnn_delegate_adaptor.cc",
+        "webnn_cpp.cpp",
```
Comment (Honry) [inline tensorflow/lite/delegates/webnn/BUILD]: We should manually copy `webnn_cpp.cpp` to this folder.

File: WORKSPACE [hunk 2]
```diff
@@ -21,3 +21,9 @@ tf_workspace1()
 load("@//tensorflow:workspace0.bzl", "tf_workspace0")
 
 tf_workspace0()
+
+new_local_repository(
+    name = "webnn_native_project",
+    path = "/home/wm/work/webml/webnn/webnn-native",
```
Comment (Honry) [inline WORKSPACE]: Hard code here. Need to find a way to reference `webnn-native`.

**Comment by Honry** [comment]
@huningxin, this is an initial PR to build webnn external delegate in native, I can now build and run models use webnn external delegate with some manual configurations.

**Build steps:** 

1. Set "WEBNN_NATIVE_DIR" env variable, via `$ export WEBNN_NATIVE_DIR=$pathToWebNNNativeDir` or passing `--action_env=WEBNN_NATIVE_DIR="$pathToWebNNNativeDir"` to bazel build command in following steps.
4. Build webnn external delegate .so file from tensorflow root directory: `$ bazel build -c opt tensorflow/lite/delegates/webnn:webnn_external_delegate.so --verbose_failures`
6. Build benchmark tool: `$ bazel build -c opt tensorflow/lite/tools/benchmark:benchmark_model`
7. Run: `$ bazel-bin/tensorflow/lite/tools/benchmark/benchmark_model   --graph=/tmp/mobilenet-v2.tflite   --external_delegate_path=bazel-bin/tensorflow/lite/delegates/webnn/webnn_external_delegate.so --external_delegate_options='webnn_device:2'`

File: tensorflow/lite/delegates/external/external_delegate.cc [hunk 3]
```diff
@@ -34,8 +34,8 @@ struct ExternalLib {
   bool load(const std::string library) {
     void* handle = SharedLibrary::LoadLibrary(library.c_str());
     if (handle == nullptr) {
-      TFLITE_LOG(TFLITE_LOG_INFO, "Unable to load external delegate from : %s",
-                 library.c_str());
+      TFLITE_LOG_PROD(tflite::TFLITE_LOG_INFO, "Unable to load external delegate from : %s",
+                 SharedLibrary::GetError());
```
Comment (huningxin) [inline tensorflow/lite/delegates/external/external_delegate.cc]: Why is this change needed?

File: third_party/webnn/webnn_native.BUILD [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    name = "webnn-native",
+    hdrs = include_files,
+    includes = ["include"],
+    strip_include_prefix = "out/Release/gen/src/include",
+    visibility = ["//visibility:public"],
+    srcs = lib_files,
+)
+
+cc_library(
+    name = "webnn-native-1",
```
Comment (huningxin) [inline third_party/webnn/webnn_native.BUILD]: why do you need both `webnn-native` and `webnn-native-1`?

File: WORKSPACE (hunk 2 above)
Comment (huningxin) [reply WORKSPACE]: It is probably OK to let user to add the local repo of webnn-native. Is there any ways to supply it via command line?

File: tensorflow/lite/delegates/webnn/BUILD (hunk 1 above)
Comment (huningxin) [reply tensorflow/lite/delegates/webnn/BUILD]: Could this file be compiled within webnn-native?

**Review by huningxin — COMMENTED** [review]
Good progress, thanks @Honry ! Left a few comments, please take a look.

File: tensorflow/lite/delegates/external/external_delegate.cc (hunk 3 above)
Comment (Honry) [reply tensorflow/lite/delegates/external/external_delegate.cc]: The original `TFLITE_LOG` will not throw error on the terminal.

File: third_party/webnn/webnn_native.BUILD (hunk 4 above)
Comment (Honry) [reply third_party/webnn/webnn_native.BUILD]: As these header files located in different directories, split them into two `cc_library` in order to distinguish different `strip_include_prefix`.

File: tensorflow/lite/delegates/webnn/BUILD (hunk 1 above)
Comment (Honry) [reply tensorflow/lite/delegates/webnn/BUILD]: Not sure, I'll consult Junwei.

File: WORKSPACE (hunk 2 above)
Comment (Honry) [reply WORKSPACE]: There's an example here https://stackoverflow.com/questions/65396733/bazel-setting-a-global-varible-from-environment-variable, I will try later.

**Comment by Honry** [comment]
@huningxin please check latest commit, I've addressed your comments.

File: tensorflow/lite/delegates/webnn/BUILD [hunk 5]
```diff
@@ -31,10 +31,26 @@ cc_library(
...
     ],
 )
 
+cc_binary(
+    name = "webnn_external_delegate.so",
+    srcs = [
+        "webnn_delegate_adaptor.cc",
+    ],
+    linkshared = 1,
+    linkstatic = 1,
```
Comment (huningxin) [inline tensorflow/lite/delegates/webnn/BUILD]: Why does it set both `linkshared` and `linkstatic`?

File: tensorflow/lite/delegates/external/external_delegate.cc (hunk 3 above)
Comment (huningxin) [reply tensorflow/lite/delegates/external/external_delegate.cc]: should it keep `library.c_str()` which is the library name?

**Review by huningxin — COMMENTED** [review]
Thanks for addressing the comments. There are two questions, please take a look.

File: tensorflow/lite/delegates/webnn/BUILD (hunk 5 above)
Comment (Honry) [reply tensorflow/lite/delegates/webnn/BUILD]: I just refer to the guideline at https://www.tensorflow.org/lite/performance/implementing_delegate#option_2_leverage_external_delegate, not quite sure the main reason why it use both here. But `linkstatic = 1` is default value, we can remove it. 

According to the bazel doc, it says: 
> If you specify both linkstatic=1 and linkshared=True, you get a single, mostly self-contained unit.

File: tensorflow/lite/delegates/external/external_delegate.cc (hunk 3 above)
Comment (Honry) [reply tensorflow/lite/delegates/external/external_delegate.cc]: `SharedLibrary::GetError()` includes the delegate path, i.e. the error may look like:

`INFO: Unable to load external delegate from : /tmp/webnn_external_delegate.so: undefined symbol: _ZN2ml7Operand12WebnnReleaseEP13MLOperandImpl`

Comment (Honry) [reply tensorflow/lite/delegates/external/external_delegate.cc]: I'll replace `TFLITE_LOG_INFO` with `TFLITE_LOG_ERROR`

**Review by huningxin — APPROVED** [review]
LGTM. thanks much @Honry !
